File: unittest/scripts/auto/js_adminapi_async/validation/locking_norecord.js
//@# INCLUDE async_utils.inc
||

//@<OUT> Create ReplicaSet another operation in progress (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Create ReplicaSet another operation in progress (fail)
Dba.createReplicaSet: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

//@<OUT> Configure instance when another operation in progress on the target (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Configure instance when another operation in progress on the target (fail)
Dba.configureReplicaSetInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>' (MYSQLSH 51500)

//@<OUT> Add instance when another operation holds an exclusive lock on the primary (fail).
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Add instance when another operation holds an exclusive lock on the primary (fail).
ReplicaSet.addInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

//@<OUT> Add instance when another operation in progress on the target (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Add instance when another operation in progress on the target (fail)
ReplicaSet.addInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>' (MYSQLSH 51500)

//@<OUT> Remove instance when another operation in progress on the target (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Remove instance when another operation in progress on the target (fail)
ReplicaSet.removeInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>' (MYSQLSH 51500)

//@<OUT> Remove instance when another operation holds an exclusive lock on the primary (fail).
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Remove instance when another operation holds an exclusive lock on the primary (fail).
ReplicaSet.removeInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

//@<OUT> Rejoin instance when another operation in progress on the target (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Rejoin instance when another operation in progress on the target (fail)
ReplicaSet.rejoinInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port2>>>' (MYSQLSH 51500)

//@<OUT> Rejoin instance when another operation holds an exclusive lock on the primary (fail).
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Rejoin instance when another operation holds an exclusive lock on the primary (fail).
ReplicaSet.rejoinInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

//@<OUT> (BUG#30673719) rejoinInstance on primary should not fail due to double self lock (using localhost).
The instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' is ONLINE and the primary of the ReplicaSet.

//@<OUT> (BUG#30673719) rejoinInstance on primary should not fail due to double self lock (using hostname).
The instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' is ONLINE and the primary of the ReplicaSet.

//@<OUT> (BUG#30673719) rejoinInstance on primary should not fail due to double self lock (using hostname_ip).
The instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' is ONLINE and the primary of the ReplicaSet.

//@<ERR> (BUG#30673719) addInstance on primary should not fail due to double self lock (using localhost).
ReplicaSet.addInstance: <<<localhost>>>:<<<__mysql_sandbox_port1>>> is already a member of this replicaset. (MYSQLSH 51301)

//@<ERR> (BUG#30673719) addInstance on primary should not fail due to double self lock (using hostname).
ReplicaSet.addInstance: <<<localhost>>>:<<<__mysql_sandbox_port1>>> is already a member of this replicaset. (MYSQLSH 51301)

//@<ERR> (BUG#30673719) addInstance on primary should not fail due to double self lock (using hostname_ip).
ReplicaSet.addInstance: <<<localhost>>>:<<<__mysql_sandbox_port1>>> is already a member of this replicaset. (MYSQLSH 51301)

//@<OUT> Set primary instance when another operation in progress on the replica set (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance 'localhost:<<<__locked_port>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Set primary instance when another operation in progress on the replica set (fail)
ReplicaSet.setPrimaryInstance: Failed to acquire lock on instance 'localhost:<<<__locked_port>>>' (MYSQLSH 51500)

//@<OUT> Force primary instance when another operation in progress on the replica set (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__locked_port>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Force primary instance when another operation in progress on the replica set (fail)
ReplicaSet.forcePrimaryInstance: Failed to acquire lock on instance '<<<localhost>>>:<<<__locked_port>>>' (MYSQLSH 51500)

//@<OUT> Remove router metadata when another operation holds an exclusive lock on the primary (fail).
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Remove router metadata when another operation holds an exclusive lock on the primary (fail).
ReplicaSet.removeRouterMetadata: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

//@<OUT> Upgrade Metadata another operation in progress (fail)
ERROR: The operation cannot be executed because it failed to acquire the lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>'. Another operation requiring access to the instance is still in progress, please wait for it to finish and try again.

//@<ERR> Upgrade Metadata another operation in progress (fail)
Dba.upgradeMetadata: Failed to acquire lock on instance '<<<localhost>>>:<<<__mysql_sandbox_port1>>>' (MYSQLSH 51500)

File: unittest/scripts/auto/py_adminapi/scripts/locking_norecord.py
#@<> utils
import mysqlsh
import threading

#@<> Setup
allowlist = "127.0.0.1," + hostname_ip;

testutil.deploy_sandbox(__mysql_sandbox_port1, 'root', {'report_host': hostname})
testutil.deploy_sandbox(__mysql_sandbox_port2, 'root', {'report_host': hostname})
testutil.snapshot_sandbox_conf(__mysql_sandbox_port1)
testutil.snapshot_sandbox_conf(__mysql_sandbox_port2)

session1 = mysql.get_session(__sandbox_uri1)
session2 = mysql.get_session(__sandbox_uri2)

shell.connect(__sandbox_uri1)

if testutil.version_check(__version, '>=', '8.0.27'):
    cluster = dba.create_cluster('cluster', {"ipAllowlist": allowlist, "gtidSetIsComplete": True, "communicationStack": "XCOM"})
else:
    cluster = dba.create_cluster('cluster', {"ipAllowlist": allowlist, "gtidSetIsComplete": True})

session1.run_sql("create schema test")
session1.run_sql("create table test.data (a int primary key auto_increment, data longtext)")

cluster.add_instance(__sandbox_uri2, {"recoveryMethod": 'incremental', "ipAllowlist": allowlist})
testutil.wait_member_transactions(__mysql_sandbox_port2, __mysql_sandbox_port1);
cluster.remove_instance(__sandbox_uri2)

for _ in range(10):
    session1.run_sql("insert into test.data values (default, repeat('x', 4*1024*1024))")

#@<> exclusive cluster lock

def add_instance(cluster, instance):
    cluster.add_instance(instance, {"recoveryMethod": 'incremental', "ipAllowlist": allowlist})

# this will make the thread block
session2.run_sql("FLUSH TABLES test.data WITH READ LOCK");

cluster_thread = dba.get_cluster();
thread = threading.Thread(target=add_instance, args=(cluster_thread, __sandbox_uri2))
thread.start()

time.sleep(3) # give time to the thead to start and block

EXPECT_THROWS(lambda: cluster.add_instance(__sandbox_uri2, {"recoveryMethod": 'incremental', "ipAllowlist": allowlist}),
    f'Failed to acquire Cluster lock through primary member \'{hostname}:{__mysql_sandbox_port1}\'')
EXPECT_STDOUT_CONTAINS(f'The operation cannot be executed because it failed to acquire the Cluster lock through primary member \'{hostname}:{__mysql_sandbox_port1}\'. Another operation requiring access to the member is still in progress, please wait for it to finish and try again.')

EXPECT_THROWS(lambda: cluster.rejoin_instance(__sandbox_uri2, {"ipAllowlist": allowlist}),
    f'Failed to acquire Cluster lock through primary member \'{hostname}:{__mysql_sandbox_port1}\'')
EXPECT_STDOUT_CONTAINS(f'The operation cannot be executed because it failed to acquire the Cluster lock through primary member \'{hostname}:{__mysql_sandbox_port1}\'. Another operation requiring access to the member is still in progress, please wait for it to finish and try again.')

session2.run_sql("UNLOCK TABLES");
thread.join()

#@<> Cleanup
cluster_thread.disconnect()
cluster.disconnect()
session.close()
session1.close()
session2.close()

testutil.destroy_sandbox(__mysql_sandbox_port1)
testutil.destroy_sandbox(__mysql_sandbox_port2)
